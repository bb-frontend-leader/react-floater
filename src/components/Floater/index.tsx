import { cloneElement, CSSProperties, isValidElement, memo, ReactNode, Ref, useMemo } from 'react';
import { PlainObject } from '@gilbarbara/types';

import { STATUS } from '../../literals';
import { CloseFunction, FloaterComponent, Statuses, Styles } from '../../types';

import Arrow from './Arrow';
import Container from './Container';

interface Props {
  arrow?: ReactNode;
  arrowRef: Ref<HTMLSpanElement>;
  component?: FloaterComponent;
  content?: ReactNode;
  floaterRef: Ref<HTMLDivElement>;
  footer?: ReactNode;
  hideArrow: boolean;
  id: string;
  onClick: CloseFunction;
  placement: string;
  positionWrapper: boolean;
  showCloseButton?: boolean;
  status: Statuses;
  styles: Styles;
  title?: ReactNode;
}

function Floater(props: Props) {
  const {
    component,
    content,
    floaterRef,
    hideArrow,
    id,
    onClick: closeFn,
    placement,
    status,
    styles,
  } = props;

  const style = useMemo(() => {
    const {
      arrow: { length },
      floater,
      floaterCentered,
      floaterClosing,
      floaterOpening,
      floaterWithComponent,
    } = styles;
    let element: CSSProperties = { ...floater };

    if (!hideArrow) {
      if (placement.startsWith('top')) {
        element.padding = `0 0 ${length}px`;
      } else if (placement.startsWith('bottom')) {
        element.padding = `${length}px 0 0`;
      } else if (placement.startsWith('left')) {
        element.padding = `0 ${length}px 0 0`;
      } else if (placement.startsWith('right')) {
        element.padding = `0 0 0 ${length}px`;
      }
    }

    if (status === STATUS.CLOSING) {
      element = { ...element, ...floaterClosing };
    }

    if (status === STATUS.OPENING || status === STATUS.OPEN) {
      element = { ...element, ...floaterOpening };
    }

    if (placement === 'center') {
      element = { ...element, ...floaterCentered };
    }

    if (component) {
      element = { ...element, ...floaterWithComponent };
    }

    return element;
  }, [component, hideArrow, placement, status, styles]);

  const shouldRender = ['closing', 'open', 'opening', 'render'].includes(status);

  const output: PlainObject<ReactNode> = {};
  const classes = ['__floater'];
  const baseProps = { role: 'tooltip' };

  if (component) {
    const componentProps = { closeFn, id, ...baseProps };

    output.content = isValidElement(component)
      ? cloneElement(component, componentProps)
      : component(componentProps);
  } else {
    output.content = <Container {...props} content={content} />;
  }

  if (status === STATUS.OPEN) {
    classes.push('__floater__open');
  }

  if (!hideArrow) {
    output.arrow = <Arrow {...props} />;
  }

  if (!shouldRender) {
    return null;
  }

  return (
    <div ref={floaterRef} id={id} style={{ zIndex: styles.options.zIndex }}>
      <div className={classes.join(' ')} style={style} {...baseProps}>
        <div className="__floater__body">
          {output.content}
          {output.arrow}
        </div>
      </div>
    </div>
  );
}

export default memo(Floater);
