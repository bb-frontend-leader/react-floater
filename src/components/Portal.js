import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { canUseDOM, isReact16 } from '../utils';

export default class ReactFloaterPortal extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([PropTypes.element, PropTypes.array]),
    hasChildren: PropTypes.bool,
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    placement: PropTypes.string,
    setRef: PropTypes.func.isRequired,
    target: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
    zIndex: PropTypes.number,
  };

  componentDidMount() {
    if (!canUseDOM()) return;

    if (!this.node) {
      this.appendNode();
    }

    if (!isReact16) {
      this.renderPortal();
    }
  }

  componentDidUpdate() {
    if (!canUseDOM()) return;

    if (!isReact16) {
      this.renderPortal();
    }
  }

  componentWillUnmount() {
    if (!canUseDOM() || !this.node) return;

    if (!isReact16) {
      ReactDOM.unmountComponentAtNode(this.node);
    }

    if (this.node && this.node.parentNode === document.body) {
      document.body.removeChild(this.node);
      this.node = undefined;
    }
  }

  appendNode() {
    const { id, zIndex } = this.props;

    if (!this.node) {
      this.node = document.createElement('div');

      /* istanbul ignore else */
      if (id) {
        this.node.id = id;
      }

      if (zIndex) {
        this.node.style.zIndex = zIndex;
      }

      document.body.appendChild(this.node);
    }
  }

  renderPortal() {
    if (!canUseDOM()) return null;

    const { children, setRef } = this.props;

    if (!this.node) {
      this.appendNode();
    }

    /* istanbul ignore else */
    if (isReact16) {
      return ReactDOM.createPortal(children, this.node);
    }

    const portal = ReactDOM.unstable_renderSubtreeIntoContainer(
      this,
      children.length > 1 ? <div>{children}</div> : children[0],
      this.node,
    );

    setRef(portal);

    return null;
  }

  renderReact16() {
    const { hasChildren, placement, target } = this.props;

    if (!hasChildren) {
      if (target || placement === 'center') {
        return this.renderPortal();
      }

      return null;
    }

    return this.renderPortal();
  }

  render() {
    if (!isReact16) {
      return null;
    }

    return this.renderReact16();
  }
}
