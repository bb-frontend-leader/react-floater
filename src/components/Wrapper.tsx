import {
  Children,
  cloneElement,
  CSSProperties,
  Fragment,
  isValidElement,
  memo,
  ReactElement,
  ReactNode,
  RefObject,
} from 'react';
import { PlainObject } from '@gilbarbara/types';
import is from 'is-lite';

import { STATUS } from '../literals';
import { useMount } from '../modules/hooks';
import { CloseFunction, Statuses } from '../types';

interface Props {
  childRef: RefObject<HTMLElement>;
  children: ReactNode;
  id: string;
  isControlled: boolean;
  onClick: CloseFunction<HTMLSpanElement>;
  onMount: () => void;
  onMouseEnter: CloseFunction<HTMLSpanElement>;
  onMouseLeave: CloseFunction<HTMLSpanElement>;
  status: Statuses;
  style?: CSSProperties;
  styles: CSSProperties;
  wrapperRef: RefObject<HTMLElement>;
}

function FloaterWrapper(props: Props) {
  const {
    childRef,
    children,
    id,
    isControlled,
    onClick,
    onMount,
    onMouseEnter,
    onMouseLeave,
    status,
    style,
    styles,
    wrapperRef,
  } = props;

  useMount(() => {
    onMount();
  });

  let element;

  const mergedStyles = {
    ...styles,
    ...style,
    ...(isValidElement(children) ? children.props.style : undefined),
  };

  let wrapperProps: PlainObject = {
    'aria-describedby': ([STATUS.OPENING, STATUS.OPEN, STATUS.CLOSING] as Statuses[]).includes(
      status,
    )
      ? id
      : undefined,
    'data-id': id,
    style: mergedStyles,
  };

  if (!isControlled) {
    wrapperProps = {
      ...wrapperProps,
      onClick,
      onMouseEnter,
      onMouseLeave,
    };
  }

  if (children) {
    if (Children.count(children) === 1 && isValidElement(children) && children.type !== Fragment) {
      // eslint-disable-next-line unicorn/prefer-ternary
      if (is.function(children.type)) {
        element = (
          <span ref={wrapperRef}>
            {cloneElement(Children.only(children) as ReactElement, {
              innerRef: childRef,
              ...wrapperProps,
            })}
          </span>
        );
      } else {
        element = cloneElement(Children.only(children) as ReactElement, {
          ref: wrapperRef,
          ...wrapperProps,
        });
      }
    } else {
      element = (
        <span ref={wrapperRef} {...wrapperProps}>
          {children}
        </span>
      );
    }
  }

  return element ?? null;
}

export default memo(FloaterWrapper);
