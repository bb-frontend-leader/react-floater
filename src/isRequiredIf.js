const VALIDATOR_ARG_ERROR_MESSAGE =
  'The typeValidator argument must be a function ' +
  'with the signature function(props, propName, componentName).';

const MESSAGE_ARG_ERROR_MESSAGE =
  'The error message is optional, but must be a string if provided.';

function propIsRequired(condition, props, propName, componentName) {
  if (typeof condition === 'boolean') {
    return condition;
  }
  if (typeof condition === 'function') {
    return condition(props, propName, componentName);
  }
  if (Boolean(condition) === true) {
    return Boolean(condition);
  }

  return false;
}

function propExists(props, propName) {
  return Object.hasOwnProperty.call(props, propName);
}

function missingPropError(props, propName, componentName, message) {
  if (message) {
    return new Error(message);
  }

  return new Error(
    `Required ${props[propName]} \`${propName}\` was not specified in \`${componentName}\`.`,
  );
}

function guardAgainstInvalidArgTypes(typeValidator, message) {
  if (typeof typeValidator !== 'function') {
    throw new TypeError(VALIDATOR_ARG_ERROR_MESSAGE);
  }

  if (Boolean(message) && typeof message !== 'string') {
    throw new TypeError(MESSAGE_ARG_ERROR_MESSAGE);
  }
}

export default function isRequiredIf(typeValidator, condition, message) {
  guardAgainstInvalidArgTypes(typeValidator, message);

  return (props, propName, componentName, ...rest) => {
    if (propIsRequired(condition, props, propName, componentName)) {
      if (propExists(props, propName)) {
        return typeValidator(props, propName, componentName, ...rest);
      }

      return missingPropError(props, propName, componentName, message);
    }

    // Is not required, so just run typeValidator.
    return typeValidator(props, propName, componentName, ...rest);
  };
}
