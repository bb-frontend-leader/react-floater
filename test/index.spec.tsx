/* eslint-disable testing-library/no-manual-cleanup */
import { ReactNode } from 'react';
import {
  act,
  cleanup,
  configure,
  fireEvent,
  render,
  RenderResult,
  screen,
} from '@testing-library/react';
import { MockInstance } from 'vitest';

import ReactFloater from '../src';
import { portalId } from '../src/modules/helpers';
import { Props } from '../src/types';

import { Button, Floaters, Styled } from './__fixtures__/components';

configure({
  testIdAttribute: 'id',
});

vi.useFakeTimers();

const mockCallback = vi.fn();
const mockGetPopper = vi.fn(() => ({ instance: {} }));

const id = 'test';
const content = 'Hello! This is my content!';

const props: Props = {
  id,
  content,
  getPopper: mockGetPopper,
};

function setup(ownProps = props, children: ReactNode = 'Places') {
  return render(<ReactFloater {...ownProps}>{children}</ReactFloater>);
}

describe('ReactFloater', () => {
  let view: RenderResult;

  const unmountView = () => {
    view?.unmount();
    cleanup();
  };

  afterEach(() => {
    mockCallback.mockClear();
  });

  describe('basic usage', () => {
    afterAll(() => {
      mockGetPopper.mockClear();
      unmountView();
    });

    it('should render the element', () => {
      view = setup(props, 'Places');

      expect(screen.getByTestId(`${id}-wrapper`)).toBeInTheDocument();
    });

    it('should render the portal, popper and floater', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));
      const portal = document.getElementById('react-floater-portal');
      const popper = portal?.firstChild;

      expect(portal).toBeInTheDocument();
      expect(portal).toHaveStyle({ zIndex: 100 });
      expect(popper).toHaveStyle({ zIndex: 100 });

      expect(screen.getByTestId('test')).toHaveClass('__floater');
      expect(screen.getByTestId('test')).not.toHaveClass('__floater__open');
    });

    it('should have called getPopper', () => {
      expect(mockGetPopper).toHaveBeenCalledWith(expect.any(Object), 'floater');
    });

    it('should show the floater after the transition ends', () => {
      fireEvent.transitionEnd(screen.getByTestId('test'));
      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should hide the floater and remove the popper', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      const portal = document.getElementById('react-floater-portal');
      const popper = portal?.firstChild;

      expect(portal).toBeInTheDocument();
      expect(popper).toBeNull();

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with multiple children', () => {
    afterAll(unmountView);

    it('should render the element', () => {
      view = setup(
        props,
        <>
          <span>Hello</span> <span>world</span>
        </>,
      );

      expect(screen.getByTestId(`${id}-wrapper`)).toBeInTheDocument();
    });

    it('should render the floater', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      expect(screen.getByTestId('test')).toBeInTheDocument();
    });

    it('should hide the floater', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with multiple instances', () => {
    afterAll(unmountView);

    it('should render the elements', () => {
      view = render(<Floaters />);

      expect(screen.getByTestId('president-wrapper')).toBeInTheDocument();
      expect(screen.getByTestId('republic-wrapper')).toBeInTheDocument();
    });

    it('should render the floaters', async () => {
      fireEvent.click(screen.getByTestId('president-wrapper'));
      fireEvent.click(screen.getByTestId('republic-wrapper'));

      expect(screen.getByTestId('president')).toHaveTextContent('It was that bearded guy!');
      expect(screen.getByTestId('republic')).toHaveTextContent('You know what I mean');
    });

    it('should hide the floaters', async () => {
      fireEvent.click(screen.getByTestId('president-wrapper'));
      fireEvent.click(screen.getByTestId('republic-wrapper'));

      fireEvent.transitionEnd(screen.getByTestId('president'));
      fireEvent.transitionEnd(screen.getByTestId('republic'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `portalElement`', () => {
    let element: HTMLDivElement;

    beforeAll(() => {
      element = document.createElement('div');
      element.id = 'floaters';

      document.body.appendChild(element);
    });

    afterAll(() => {
      unmountView();
      document.body.removeChild(element);
    });

    it('should render in the "portalElement"', async () => {
      view = render(<Floaters portalElement={element} />);

      expect(screen.queryByTestId(portalId)).not.toBeInTheDocument();
      expect(screen.getByTestId('floaters')).toBeInTheDocument();

      fireEvent.click(screen.getByTestId('president-wrapper'));
      fireEvent.click(screen.getByTestId('republic-wrapper'));

      expect(screen.getByTestId('president')).toHaveTextContent('It was that bearded guy!');
      expect(screen.getByTestId('republic')).toHaveTextContent('You know what I mean');
    });
  });

  describe('with `autoOpen`', () => {
    afterAll(unmountView);

    it('should have rendered the Floater initially open', async () => {
      view = setup({
        ...props,
        autoOpen: true,
      });

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));
      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should hide the floater', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });

    it('should show the floater again', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });
  });

  describe('with `callback`', () => {
    afterAll(unmountView);

    it('should call the callback function on open', async () => {
      view = setup({
        ...props,
        callback: mockCallback,
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(mockCallback).toHaveBeenCalledWith('open', {
        autoOpen: false,
        callback: mockCallback,
        children: 'Places',
        content: 'Hello! This is my content!',
        debug: false,
        disableFlip: false,
        disableHoverToClick: false,
        event: 'click',
        eventDelay: 0.4,
        getPopper: mockGetPopper,
        hideArrow: false,
        id,
        offset: 15,
        placement: 'bottom',
        showCloseButton: false,
        styles: {},
        target: null,
        wrapperOptions: { position: false },
      });
    });

    it('should call the callback function on close', async () => {
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(mockCallback).toHaveBeenCalledWith('close', {
        autoOpen: false,
        callback: mockCallback,
        children: 'Places',
        content: 'Hello! This is my content!',
        debug: false,
        disableFlip: false,
        disableHoverToClick: false,
        event: 'click',
        eventDelay: 0.4,
        getPopper: mockGetPopper,
        hideArrow: false,
        id,
        offset: 15,
        placement: 'bottom',
        showCloseButton: false,
        styles: {},
        target: null,
        wrapperOptions: { position: false },
      });
    });
  });

  describe('with `debug`', () => {
    let consoleGroupCollapsed: MockInstance;
    let consoleLog: MockInstance;

    beforeAll(() => {
      consoleGroupCollapsed = vi
        .spyOn(console, 'groupCollapsed')
        .mockImplementation(() => undefined);
      consoleLog = vi.spyOn(console, 'log').mockImplementation(() => undefined);
    });

    afterAll(() => {
      consoleGroupCollapsed.mockRestore();
      consoleLog.mockRestore();
      unmountView();
    });

    it('should be able to show and hide the floater', async () => {
      view = setup({
        ...props,
        debug: true,
      });

      expect(consoleGroupCollapsed).toHaveBeenCalledWith(
        '%creact-floater: init',
        'color: #9b00ff; font-weight: bold; font-size: 12px;',
      );

      expect(consoleLog).toHaveBeenCalledWith(
        expect.objectContaining({
          floater: null,
          hasChildren: true,
          hasTarget: false,
          isControlled: false,
          positionWrapper: false,
          target: expect.anything(),
        }),
      );

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      expect(consoleGroupCollapsed).toHaveBeenCalledWith(
        '%creact-floater: click',
        'color: #9b00ff; font-weight: bold; font-size: 12px;',
      );
      expect(consoleLog).toHaveBeenCalledWith({ event: 'click', status: 'opening' });

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');

      expect(consoleGroupCollapsed).toHaveBeenCalledTimes(2);
      expect(consoleLog).toHaveBeenCalledTimes(2);
    });
  });

  describe('with `event` hover', () => {
    afterAll(unmountView);

    it('should be able to show the floater', async () => {
      view = setup({
        ...props,
        event: 'hover',
        disableHoverToClick: true,
      });

      fireEvent.mouseEnter(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should close itself after `eventDelay`', async () => {
      fireEvent.mouseLeave(screen.getByTestId(`${id}-wrapper`));

      act(() => {
        vi.advanceTimersByTime(1000);
      });

      expect(screen.getByTestId('test')).not.toHaveClass('__floater__open');

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `event` hover and `eventDelay` set to 0', () => {
    afterAll(unmountView);

    it('should be able to show the floater', async () => {
      view = setup({
        ...props,
        event: 'hover',
        eventDelay: 0,
      });

      fireEvent.mouseEnter(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should have close itself immediately', async () => {
      fireEvent.mouseLeave(screen.getByTestId(`${id}-wrapper`));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `title`, `footer` and `closeBtn`', () => {
    afterAll(unmountView);

    it('should render the floater', async () => {
      view = setup({
        ...props,
        footer: 'Footer',
        showCloseButton: true,
        title: 'Title',
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });
  });

  describe('with `open`', () => {
    const localProps = {
      ...props,
      open: false,
    };

    afterAll(unmountView);

    it('should not be able to show the floater with click', async () => {
      view = setup(localProps);

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      expect(screen.queryByTestId('test')).not.toBeInTheDocument();
    });

    it('should not be able to show the floater with hover', async () => {
      view = setup({ ...localProps, event: 'hover' });

      fireEvent.mouseEnter(screen.getAllByTestId(`${id}-wrapper`)[0]);

      expect(screen.queryByTestId('test')).not.toBeInTheDocument();
    });

    it('should show the floater when `open` is true', async () => {
      view.rerender(
        <ReactFloater {...localProps} open>
          {content}
        </ReactFloater>,
      );

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('test')).toHaveClass('__floater__open');
    });

    it('should close the floater when `open` is false', async () => {
      view.rerender(
        <ReactFloater {...localProps} open={false}>
          {content}
        </ReactFloater>,
      );

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.queryByTestId('test')).not.toBeInTheDocument();
    });
  });

  describe('with `component` as function', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup(
        {
          ...props,
          content: undefined,
          component: Styled,
        },
        <Button />,
      );

      fireEvent.click(screen.getByTestId('button'));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });

    it('should close the floater with `closeFn` prop', async () => {
      fireEvent.click(screen.getByRole('button', { name: /close/i }));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `component` as element', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup(
        {
          ...props,
          content: undefined,
          component: <Styled />,
          styles: {
            options: {
              zIndex: 1000,
            },
          },
        },
        <span>Places</span>,
      );

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });

    it('should close the floater with `closeFn` prop', async () => {
      fireEvent.click(screen.getByRole('button', { name: /close/i }));

      fireEvent.transitionEnd(screen.getByTestId('test'));

      expect(screen.getByTestId('react-floater-portal')).toBeEmptyDOMElement();
    });
  });

  describe('with `placement` top', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup({
        ...props,
        placement: 'top',
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });
  });

  describe('with `placement` left', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup({
        ...props,
        placement: 'left',
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });
  });

  describe('with `placement` right', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup({
        ...props,
        placement: 'right',
      });

      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId('test')).toMatchSnapshot();
    });
  });

  describe('with `placement` center', () => {
    afterAll(unmountView);

    it('should show the floater with click', async () => {
      view = setup({
        ...props,
        placement: 'center',
      });
      fireEvent.click(screen.getByTestId(`${id}-wrapper`));

      await act(async () => {
        vi.runOnlyPendingTimers();
      });

      expect(screen.getByTestId(id)).toMatchSnapshot();
    });
  });

  describe('with `wrapperOptions`', () => {
    afterAll(unmountView);

    it('should render properly', async () => {
      view = render(
        <>
          <span className="external" id="external">
            external
          </span>
          <ReactFloater
            content={
              <div style={{ fontSize: 32 }}>Yeah, this is how we use to look back in the day!</div>
            }
            disableFlip
            id={id}
            placement="top"
            target=".external"
            wrapperOptions={{
              offset: -20,
              placement: 'bottom',
              position: true,
            }}
          >
            <span id="beacon">Beacon</span>
          </ReactFloater>
        </>,
      );

      fireEvent.click(screen.getByTestId('test-wrapper'));

      expect(screen.getByTestId(id)).toHaveTextContent(
        'Yeah, this is how we use to look back in the day!',
      );
    });
  });
});
