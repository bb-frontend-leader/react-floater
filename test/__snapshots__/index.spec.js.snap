// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactFloater with \`component\` as element should have rendered the component 1`] = `
<div
  className="__floater__body"
>
  <Styled
    closeFn={[Function]}
  >
    <styled.div>
      <div
        className="sc-bdfCDU kJyuGD"
      >
        <div>
          Styled
        </div>
        <button
          onClick={[Function]}
          type="button"
        >
          close
        </button>
      </div>
    </styled.div>
  </Styled>
  <FloaterArrow
    component={<Styled />}
    disableAnimation={false}
    handleClick={[Function]}
    hideArrow={false}
    placement="bottom"
    positionWrapper={false}
    setArrowRef={[Function]}
    setFloaterRef={[Function]}
    showCloseButton={false}
    status="open"
    styles={
      {
        "arrow": {
          "color": "#fff",
          "display": "inline-flex",
          "length": 16,
          "margin": 8,
          "position": "absolute",
          "spread": 32,
        },
        "close": {
          "WebkitAppearance": "none",
          "backgroundColor": "transparent",
          "border": 0,
          "borderRadius": 0,
          "color": "#555",
          "fontSize": 0,
          "height": 15,
          "outline": "none",
          "padding": 10,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 15,
        },
        "container": {
          "backgroundColor": "#fff",
          "color": "#666",
          "minHeight": 60,
          "minWidth": 200,
          "padding": 20,
          "position": "relative",
          "zIndex": 10,
        },
        "content": {
          "fontSize": 15,
        },
        "floater": {
          "display": "inline-block",
          "filter": "drop-shadow(0 0 3px rgba(0, 0, 0, 0.3))",
          "maxWidth": 300,
          "opacity": 0,
          "position": "relative",
          "transition": "opacity 0.3s",
          "visibility": "hidden",
          "zIndex": 1000,
        },
        "floaterCentered": {
          "left": "50%",
          "position": "fixed",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        },
        "floaterClosing": {
          "opacity": 0,
          "visibility": "visible",
        },
        "floaterOpening": {
          "opacity": 1,
          "visibility": "visible",
        },
        "floaterWithAnimation": {
          "opacity": 1,
          "transition": "opacity 0.3s, transform 0.2s",
          "visibility": "visible",
        },
        "floaterWithComponent": {
          "maxWidth": "100%",
        },
        "footer": {
          "borderTop": "1px solid #ccc",
          "fontSize": 13,
          "marginTop": 10,
          "paddingTop": 5,
        },
        "options": {
          "zIndex": 1000,
        },
        "title": {
          "borderBottom": "1px solid #555",
          "color": "#555",
          "fontSize": 18,
          "marginBottom": 5,
          "paddingBottom": 6,
          "paddingRight": 18,
        },
        "wrapper": {
          "bottom": "",
          "cursor": "help",
          "display": "inline-flex",
          "flexDirection": "column",
          "left": "",
          "position": "",
          "right": "",
          "top": "",
          "zIndex": 1000,
        },
        "wrapperPosition": {
          "left": -1000,
          "position": "absolute",
          "top": -1000,
          "visibility": "hidden",
        },
      }
    }
  >
    <div
      className="__floater__arrow"
      style={
        {
          "height": 16,
          "left": 0,
          "pointerEvents": "none",
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": "100%",
        }
      }
    >
      <span
        style={
          {
            "display": "inline-flex",
            "marginLeft": 8,
            "marginRight": 8,
            "position": "absolute",
            "top": 0,
          }
        }
      >
        <svg
          height={16}
          version="1.1"
          width={32}
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            fill="#fff"
            points="32,16 16,0 0,16"
          />
        </svg>
      </span>
    </div>
  </FloaterArrow>
</div>
`;

exports[`ReactFloater with \`component\` as function should have rendered the component 1`] = `
<div
  className="__floater__body"
>
  <styled.div>
    <div
      className="sc-bdfCDU kJyuGD"
    >
      <div>
        Styled
      </div>
      <button
        onClick={[Function]}
        type="button"
      >
        close
      </button>
    </div>
  </styled.div>
  <FloaterArrow
    component={[Function]}
    disableAnimation={false}
    handleClick={[Function]}
    hideArrow={false}
    placement="bottom"
    positionWrapper={false}
    setArrowRef={[Function]}
    setFloaterRef={[Function]}
    showCloseButton={false}
    status="open"
    styles={
      {
        "arrow": {
          "color": "#fff",
          "display": "inline-flex",
          "length": 16,
          "margin": 8,
          "position": "absolute",
          "spread": 32,
        },
        "close": {
          "WebkitAppearance": "none",
          "backgroundColor": "transparent",
          "border": 0,
          "borderRadius": 0,
          "color": "#555",
          "fontSize": 0,
          "height": 15,
          "outline": "none",
          "padding": 10,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 15,
        },
        "container": {
          "backgroundColor": "#fff",
          "color": "#666",
          "minHeight": 60,
          "minWidth": 200,
          "padding": 20,
          "position": "relative",
          "zIndex": 10,
        },
        "content": {
          "fontSize": 15,
        },
        "floater": {
          "display": "inline-block",
          "filter": "drop-shadow(0 0 3px rgba(0, 0, 0, 0.3))",
          "maxWidth": 300,
          "opacity": 0,
          "position": "relative",
          "transition": "opacity 0.3s",
          "visibility": "hidden",
          "zIndex": 100,
        },
        "floaterCentered": {
          "left": "50%",
          "position": "fixed",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        },
        "floaterClosing": {
          "opacity": 0,
          "visibility": "visible",
        },
        "floaterOpening": {
          "opacity": 1,
          "visibility": "visible",
        },
        "floaterWithAnimation": {
          "opacity": 1,
          "transition": "opacity 0.3s, transform 0.2s",
          "visibility": "visible",
        },
        "floaterWithComponent": {
          "maxWidth": "100%",
        },
        "footer": {
          "borderTop": "1px solid #ccc",
          "fontSize": 13,
          "marginTop": 10,
          "paddingTop": 5,
        },
        "options": {
          "zIndex": 100,
        },
        "title": {
          "borderBottom": "1px solid #555",
          "color": "#555",
          "fontSize": 18,
          "marginBottom": 5,
          "paddingBottom": 6,
          "paddingRight": 18,
        },
        "wrapper": {
          "bottom": "",
          "cursor": "help",
          "display": "inline-flex",
          "flexDirection": "column",
          "left": "",
          "position": "",
          "right": "",
          "top": "",
          "zIndex": 100,
        },
        "wrapperPosition": {
          "left": -1000,
          "position": "absolute",
          "top": -1000,
          "visibility": "hidden",
        },
      }
    }
  >
    <div
      className="__floater__arrow"
      style={
        {
          "height": 16,
          "left": 0,
          "pointerEvents": "none",
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": "100%",
        }
      }
    >
      <span
        style={
          {
            "display": "inline-flex",
            "marginLeft": 8,
            "marginRight": 8,
            "position": "absolute",
            "top": 0,
          }
        }
      >
        <svg
          height={16}
          version="1.1"
          width={32}
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            fill="#fff"
            points="32,16 16,0 0,16"
          />
        </svg>
      </span>
    </div>
  </FloaterArrow>
</div>
`;
